### PR TITLE
Refactoring Module/Task/ModuleBuilder

### PR DESCRIPTION
**Pull request by shinae-woo**
This PR doesn't change any functionality, but refactor the existing `module.h` and `module.cc` files as they are integrating too much functionality in a file. Once these changes are approved, then I can make change the internal scheduling logic.

Major changes are
- Get global info about `module_graphs` out of `ModuleBuilder` class
- Get `Task/ModuleTask` out of `module.h`

### CODE REVIEW

File: core/module.h [hunk 1]
```diff
@@ -192,14 +183,44 @@ class ModuleBuilder {
...
+  const Commands cmds_;
+  const module_init_func_t init_func_;
+};
+
+// A static class for managing module graphs
+class ModuleGraph {
+ public:
+
+  // Return true if any module from the builder exists
+  static bool ExistModuleClass(const ModuleBuilder *);
```
Comment (chris3torek) [inline core/module.h]: nit: `HasModuleOfClass` might be a better name (although I'm not sure I really like this name either—the tricky bit is that we're asking whether any module exists using the class _implied by_ the given `ModuleBuilder` instance, which is hard to encode in the predicate name...).

File: core/bessctl.cc [hunk 2]
```diff
@@ -1241,10 +1241,10 @@ class BESSControlImpl final : public BESSControl::Service {
 
     if (is_any_worker_running()) {
       propagate_active_worker();
-      if (m1->num_active_workers() || m2->num_active_workers() ) {
-	WorkerPauser wp; // Only pause when absolutely required
-	ret = m1->ConnectModules(ogate, m2, igate);
-	goto done;
+      if (m1->num_active_workers() || m2->num_active_workers()) {
+        WorkerPauser wp;  // Only pause when absolutely required
```
Comment (chris3torek) [inline core/bessctl.cc]: FYI I think this reformat might be the same as one in one of @melvinw PRs, hopefully they match exactly and Git will be able to handle it easily 😺

**Comment by barath** [comment]
Will look.

File: core/module.h (hunk 1 above)
Comment (shinae-woo) [reply core/module.h]: Really, it is better!

File: core/module.h [hunk 3]
```diff
@@ -132,7 +97,8 @@ struct Command {
 
 using Commands = std::vector<struct Command>;
 
-// A class for generating new Modules of a particular type.
+// A class for managing modules of 'a particular type'.
+// It creates new modules and forwarding module-specific commands
```
Comment (barath) [inline core/module.h]: Minor: perhaps rewrite as `Creates new modules and forwards module-specific commands.`

File: core/module.h [hunk 4]
```diff
@@ -150,34 +116,24 @@ class ModuleBuilder {
...
                                   module_init_func_t init_func);
-
   static bool DeregisterModuleClass(const std::string &class_name);
 
   static std::map<std::string, ModuleBuilder> &all_module_builders_holder(
       bool reset = false);
   static const std::map<std::string, ModuleBuilder> &all_module_builders();
 
-  static const std::map<std::string, Module *> &all_modules();
+  /* returns a pointer to the created module.
```
Comment (barath) [inline core/module.h]: Perhaps update the comment?

File: core/module_graph.cc [hunk 5]
```diff
@@ -0,0 +1,254 @@
+// Copyright (c) 2014-2016, The Regents of the University of California.
```
Comment (barath) [inline core/module_graph.cc]: I think this copyright line now should be 2014-2017.

File: core/module_graph.cc [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+
+#include <glog/logging.h>
+
+#include "module.h"
+
+std::map<std::string, Module *> ModuleGraph::all_modules_;
+std::unordered_map<std::string, Node> ModuleGraph::module_graph_;
+std::unordered_set<std::string> ModuleGraph::tasks_;
+
+// Finds the next module that implements a task, and updates it's parents
```
Comment (barath) [inline core/module_graph.cc]: Minor: we can omit comments in the .cc file if it's also in the .h

File: core/module_graph.cc [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+  if (erase) {
+    all_modules_.erase(m->name());
+  }
+
+  module_graph_.erase(m->name());
+  if (m->is_task_) {
+    tasks_.erase(m->name());
+  }
+
+  delete m;
```
Comment (barath) [inline core/module_graph.cc]: `m` appears to only get deleted here; what happens if one of the other execution paths is taken?  If we need to delete in those cases too, maybe we can put m into a `std::unique_ptr` at the top?

File: core/module_graph.h [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+  // Disconnects two modules (`to` and `from`) together in `module_graph_`.
+  static bool RemoveEdge(const std::string &from, const std::string &to);
+
+ private:
+  // Updates the parents of modules with tasks by traversing `module_graph_` and
+  // ignoring all modules that are not tasks.
+  static bool UpdateTaskGraph();
+
+  // Finds the next module that implements a task, and updates it's parents
```
Comment (barath) [inline core/module_graph.h]: I'm not sure I understand from this comment what this method does.  (I know the comment was already there, just figured we might clarify it.)

Minor: its

File: core/module_graph.h [hunk 9]
```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2014-2016, The Regents of the University of California.
```
Comment (barath) [inline core/module_graph.h]: 2014-2017

File: core/module_graph.h [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+  // Module that this Node represents.
+  Module *module_;
+
+  // Children of `module_` in the pipeline.
+  std::unordered_set<std::string> children_;
+
+  DISALLOW_COPY_AND_ASSIGN(Node);
+};
+
+class ModuleGraph {
```
Comment (barath) [inline core/module_graph.h]: Could you add a class comment?

File: core/task.cc [hunk 11]
```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2014-2016, The Regents of the University of California.
```
Comment (barath) [inline core/task.cc]: 2014-2017

File: core/task.h [hunk 12]
```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2014-2016, The Regents of the University of California.
```
Comment (barath) [inline core/task.h]: 2014-2017

**Comment by barath** [comment]
LGTM.  Just a few minor things.

File: core/module_graph.cc (hunk 7 above)
Comment (shinae-woo) [reply core/module_graph.cc]: Yes, definitely it would be better if we make the ownership of modules (as well as TCs, tasks) much clearer.

**Comment by shinae-woo** [comment]
Thank you for the review, @barath. Mostly addressed. The ownership parts can be addressed separately so that makes changes gradually. 

@ddiproietto, Thank you for confirming the performance. I often found that compilers are much smarter and more efficient than myself 😢...

File: core/module_graph.cc (hunk 7 above)
Comment (shinae-woo) [reply core/module_graph.cc]: I am creating a new PR to address the issues handling module creation and deletion correctly. Not only with `DestroyModule()`, but also between `CreateModule()` and `AddModule()` has the same problem.

**Comment by sangjinhan** [comment]
Merged, thanks!

tumblr_m4pxpd98cr1r6jd7fo1_1280

**Pull request merged**